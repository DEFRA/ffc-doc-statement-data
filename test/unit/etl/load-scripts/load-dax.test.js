const db = require('../../../../app/data')
const { loadDAX } = require('../../../../app/etl/load-scripts/load-dax')

jest.mock('../../../../app/data', () => ({
  sequelize: {
    query: jest.fn()
  }
}))

describe('loadDAX', () => {
  const startDate = '2023-01-01'
  const transaction = {}

  beforeEach(() => {
    db.sequelize.query.mockClear()
  })

  test('should call sequelize.query with correct SQL and parameters', async () => {
    await loadDAX(startDate, transaction)

    expect(db.sequelize.query).toHaveBeenCalledWith(`
    INSERT INTO dax (
      "paymentReference", "calculationId", "paymentPeriod",
      "paymentAmount", "transactionDate"
    )
    SELECT
      T.payment_ref AS paymentReference,
      T.calculation_id AS calculationId,
      T.quarter AS paymentPeriod, 
      T.total_amount AS paymentAmount,
      T.transdate AS transactionDate 
    FROM etl_interm_total T
    WHERE T.etl_inserted_dt > :startDate;
  `, {
      replacements: {
        startDate
      },
      raw: true,
      transaction
    })
  })

  test('should handle errors thrown by sequelize.query', async () => {
    db.sequelize.query.mockRejectedValue(new Error('Query failed'))

    await expect(loadDAX(startDate, transaction)).rejects.toThrow('Query failed')
  })
})
