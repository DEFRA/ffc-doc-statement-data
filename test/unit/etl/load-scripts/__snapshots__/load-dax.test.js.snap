// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadDAX should call sequelize.query with correct SQL and parameters 1`] = `
[MockFunction] {
  "calls": [
    [
      "
    INSERT INTO public.dax (
      "paymentReference", "calculationId", "paymentPeriod",
      "paymentAmount", "transactionDate"
    )
    SELECT DISTINCT
      T."paymentRef" AS "paymentReference",
      T."calculationId" AS "calculationId",
      T.quarter AS "paymentPeriod", 
      T."totalAmount" AS "paymentAmount",
      T.transdate AS "transactionDate" 
    FROM public."etlIntermTotal" T
    LEFT JOIN public."delinkedCalculation" D ON T."calculationId" = D."calculationId"
    WHERE T."etlInsertedDt" > :startDate
      AND D."calculationId" IS NULL
    ON CONFLICT ("paymentReference", "calculationId")
    DO UPDATE SET
      "paymentAmount" = EXCLUDED."paymentAmount",
      "datePublished" = NULL;
;
  ",
      {
        "raw": true,
        "replacements": {
          "startDate": "2023-01-01",
        },
        "transaction": {},
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
